### PR TITLE
Topology aware service routing

### LINKED ISSUE #58744: Support topology aware service routing
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

/sig network

/assign

**What happened**:

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by m1093782566**
**What this PR does / why we need it**:
Support topology aware service routing - allow users reach the service endpoints in the same topology domain, for example, same host, same rack, same region, same zone and so on... After all, topology is arbitrary.

**Which issue(s) this PR fixes**:
Fixes #58744

**Special notes for your reviewer**:

Design proposal, see: kubernetes/community#1551

**Release note**:

```release-note
Support topology aware service routing
```

/sig network

### CODE REVIEW

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -201,6 +236,62 @@ func (e *EndpointController) addPod(obj interface{}) {
...
+}
+
+func (e *EndpointController) updateNode(old, cur interface{}) {
+	// We only need to update endpoint address when node labels changed.
+	oldNode := old.(*v1.Node)
+	curNode := cur.(*v1.Node)
+	if apiequality.Semantic.DeepEqual(oldNode.Labels, curNode.Labels) {
+		return
+	}
+	pods, err := e.nodePodsCache.ListPodsOnNode(curNode.Name)
```
Comment (hzxuzhonghu) [inline pkg/controller/endpoint/endpoints_controller.go]: `pods` maybe data race.

Comment (m1093782566) [reply pkg/controller/endpoint/endpoints_controller.go]: Please check `ListPodsOnNode()` in cache.go

```
func (cache *NodePodsCache) ListPodsOnNode(node string) ([]*v1.Pod, error) {
	cache.mu.Lock()
	defer cache.mu.Unlock()

	return cache.NodePods[node], nil
}
```

We protect the `nodePods` cache via mutex.

**Comment by thockin** [comment]
I just got back in town, will look at this ASAP. The big questions I have:

Should this be part of Service?  It was raised on the last call that this makes a lot of sense outside of Service in a more admin-owned way.  If we did that, it starts to look a LOT like Istio - a separation of Services and Policies.  I think Istio is a great thing, and will (eventually) be the basis for a lot of higher-level functionality, but I am sensitive to the idea that it may be premature.

That said, if Istio is the likely better answer, should we extend core Services in this way, knowing that it sort of overrides things that properly belong in something like Istio?  Or should we try to converge with Istio concepts in a way that we can extend into Istio later?  But Istio can't become required...

**Comment by m1093782566** [comment]
Although Istio is a great thing, I agree with you that Istio should not become required(at least for now). Because currently Istio is not a  must-be-selected option to many kubernetes users and they still need a kubernetes-core/lower-level way to resolve service topology problems. For example, as @smarterclayton mentioned,  

> In container identity wg, being able to give daemonset pods a unique identity per host is on the 2018 plan, and ensuring local pods can communicate to local node services securely is a key goal there.

Regards admin-owned VS. people-oriented way. I think we should give people(users) the right to configure topology policies, just like what we have done for pods in scheduling area.  Furthermore, we may need to consider the multi-tenancy scenario in the future - tenants should have the permission to configure their own service policy?

**Comment by cmluciano** [comment]
@kubernetes/sig-network-pr-reviews

**Comment by thockin** [comment]
I don't think we can proceed with this PR until we get consensus on the proposal.

**Comment by thockin** [comment]
@johnbelamaric 

/lifecycle frozen
/remove-lifecycle stale

**Comment by johnbelamaric** [comment]
There has been considerable effort put into the modeling of topology in sig-storage for https://github.com/kubernetes/features/issues/561. I am going to take a look at how that affects this proposal and PR. @m1093782566 are you still pursuing this?

**Comment by thockin** [comment]
Should we close this now that there are other PRs in flight?

**Comment by m1093782566** [comment]
Okay, closing this PR in favor of #72046.

**Pull request closed without merging**